storagenode/storagenodedb/database.go

```diff
@@ -290,24 +332,57 @@ func (db *DB) RoutingTable() (kdb, ndb, adb storage.KeyValueStore) {
 
 // RawDatabases are required for testing purposes
 func (db *DB) RawDatabases() map[string]SQLDB {
-	return map[string]SQLDB{
-		BandwidthDBName:       db.bandwidthDB,
-		OrdersDBName:          db.ordersDB,
-		PieceExpirationDBName: db.pieceExpirationDB,
-		PieceSpaceUsedDBName:  db.pieceSpaceUsedDB,
-		ReputationDBName:      db.reputationDB,
-		StorageUsageDBName:    db.storageUsageDB,
-		UsedSerialsDBName:     db.usedSerialsDB,
-		PieceInfoDBName:       db.v0PieceInfoDB,
-		DeprecatedInfoDBName:  db.deprecatedInfoDB,
-		SatellitesDBName:      db.satellitesDB,
+	return db.sqlDatabases
+}
+
+// migrateToDB is a helper method that performs the migration from the
+// deprecatedInfoDB to the specified new db. It first closes and deletes any
+// existing database to guarantee idempotence. After migration it also closes
+// and re-opens the new database to allow the system to recover used disk space.
+func (db *DB) migrateToDB(ctx context.Context, dbName string, tablesToKeep ...string) error {
+	err := db.closeDatabase(dbName)
+	if err != nil {
+		return ErrDatabase.Wrap(err)
+	}
+
+	path := db.filepathFromDBName(dbName)
+
+	if _, err := os.Stat(path); err == nil {
+		err = os.Remove(path)
+		if err != nil {
+			return ErrDatabase.Wrap(err)
+		}
+	}
+
+	err = db.openDatabase(dbName)
+	if err != nil {
+		return ErrDatabase.Wrap(err)
+	}
+
+	err = sqliteutil.MigrateTablesToDatabase(ctx, db.rawDatabaseFromName(DeprecatedInfoDBName), db.rawDatabaseFromName(dbName), tablesToKeep...)
+	if err != nil {
+		return ErrDatabase.Wrap(err)
 	}
+
+	// We need to close and re-open the database we have just migrated *to* in
+	// order to recover any excess disk usage that was freed in the VACUUM call
+	err = db.closeDatabase(dbName)
+	if err != nil {
+		return ErrDatabase.Wrap(err)
+	}
+
+	err = db.openDatabase(dbName)
+	if err != nil {
+		return ErrDatabase.Wrap(err)
+	}
+
+	return nil
 }
 
 // Migration returns table migrations.
-func (db *DB) Migration() *migrate.Migration {
+func (db *DB) Migration(ctx context.Context) *migrate.Migration {
```

Comment:
One of the reasons the migration didn't have a context, was that probably we don't want people cancelling the migration mid-way-through... altough when the migration code is idempotent, then there shouldn't be any problems.